Zend/zend_exceptions.c

```diff
@@ -307,28 +311,30 @@ ZEND_METHOD(Exception, __construct)
 {
 	zend_string *message = NULL;
 	zend_long   code = 0;
-	zval  tmp, *object, *previous = NULL;
-	zend_class_entry *base_ce;
+	zval  *object, *previous = NULL;
 
 	object = ZEND_THIS;
-	base_ce = i_get_exception_base(Z_OBJ_P(object));
 
 	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|SlO!", &message, &code, &previous, zend_ce_throwable) == FAILURE) {
 		RETURN_THROWS();
 	}
 
 	if (message) {
-		ZVAL_STR(&tmp, message);
-		zend_update_property_ex(base_ce, Z_OBJ_P(object), ZSTR_KNOWN(ZEND_STR_MESSAGE), &tmp);
+		zval *tmp = zend_obj_prop_num_checked(Z_OBJ_P(object), 0, ZSTR_KNOWN(ZEND_STR_MESSAGE));
+		zval_ptr_dtor(tmp);
+		ZVAL_STR_COPY(tmp, message);
 	}
 
 	if (code) {
-		ZVAL_LONG(&tmp, code);
-		zend_update_property_ex(base_ce, Z_OBJ_P(object), ZSTR_KNOWN(ZEND_STR_CODE), &tmp);
+		zval *tmp = zend_obj_prop_num_checked(Z_OBJ_P(object), 2, ZSTR_KNOWN(ZEND_STR_CODE));
+		zval_ptr_dtor(tmp);
+		ZVAL_LONG(tmp, code);
 	}
 
 	if (previous) {
-		zend_update_property_ex(base_ce, Z_OBJ_P(object), ZSTR_KNOWN(ZEND_STR_PREVIOUS), previous);
+		zval *tmp = zend_obj_prop_num_checked(Z_OBJ_P(object), 6, ZSTR_KNOWN(ZEND_STR_PREVIOUS));
+		zval_ptr_dtor(tmp);
+		ZVAL_COPY(tmp, previous);
```

Comment:
You sadly have to do the garbage-move dance as you cannot rely on this being called before any object would have been assigned to these values - at least those which can contain objects. I.e. previous in Exception and ErrorException.

Reply:
Right thanks! I'll use `zval_ptr_safe_dtor` though.

Reply:
Right, that works as the property is nullable.